### PR TITLE
Settings page - error when saving fails(Last System Access threshold value)

### PR DESCRIPTION
**Pull request by rohitvinnakota**
Adds a d2l-alert-toast message when trying to save and close an invalid set of settings. Currently, the message works for the last system access threshold value case but the functionality can be extended.

[US122868](https://rally1.rallydev.com/#/23525809118d/workviews/collection?detail=%2Fuserstory%2F459991179776%2Ftasks&itemRef=%2Fhierarchicalrequirement%2F459991179776&previousTokens=%2F23525809118d%2Fcustom%2F12155517225%3F&project=%2Fproject%2F23525809118&ref=%2Fhierarchicalrequirement%2F459991179776&relationship=tasks&fdp=true?fdp=true): [Engagement][Config] Settings page - error when saving fails

image
image

(The value in the screenshot's input is 233, not 23)

Screen reader: Clicking Save and close on an invalid system access value immediately triggers the alert and reads it out.
Responsiveness: Alert centers around screen when window is resized. 
Tested on edge, chrome, firefox

Security: n/a
Devops: n/a

- [x] Demo to Kevin

fyi: @johngwilkinson @hyehayes @NicholasHallman @anhill-D2L @MykolaGalian

### CODE REVIEW

File: components/custom-toast-message.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+		`;
+	}
+
+	systemLastAccessError() {
+		this.shadowRoot.querySelector('d2l-alert-toast').innerHTML = this.localize('settings:invalidSystemAccessValueToast');
+		this.shadowRoot.querySelector('d2l-alert-toast').open = true;
+	}
+}
+
+customElements.define('d2l-insights-custom-toast-message', CustomToastMessage);
```
Comment (Vitalii-Misechko) [inline components/custom-toast-message.js:20]: `thought` This component does not provide any additional behavior on top of `d2l-alert-toast`. Should we delete it?

File: components/dashboard-settings.js [hunk 2]
```diff
@@ -249,6 +250,11 @@ class DashboardSettings extends RtlMixin(Localizer(LitElement)) {
 			includeRoles: this._selectedRoleIds
 		};
 
+		if (cardSelectionList.isInvalidSystemAccessValue()) {
+			this.shadowRoot.querySelector('d2l-insights-custom-toast-message').systemLastAccessError();
+			return;
```
Comment (Vitalii-Misechko) [inline components/dashboard-settings.js:255]: Have you considered introducing `open` property into `d2l-insights-custom-toast-message` (and `_saveError` into this class)? This way you could
```JS
<d2l-insights-custom-toast-message ?open="${this._saveError}">
...
if (cardSelectionList.isInvalidSystemAccessValue()) {
  this._saveError= true
  return
}
```
I think this way is more natural for the Lit-Element framework.

File: test/components/custom-toast-message.test.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+			const el = await fixture(html`<d2l-insights-custom-toast-message></d2l-insights-custom-toast-message>`);
+			await expect(el).to.be.accessible();
+		});
+	});
+
+	describe('render', () => {
+		it('should render the invalid system last accessed value toast message as expected', async() => {
+			const el = await fixture(html`<d2l-insights-custom-toast-message></d2l-insights-custom-toast-message>`);
+			await new Promise(resolve => setTimeout(resolve, 200)); // allow fetch to run
```
Comment (Vitalii-Misechko) [inline test/components/custom-toast-message.test.js:24]: `nit` wrong comment

File: test/components/custom-toast-message.test.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+			expect(el.shadowRoot.querySelector('d2l-alert-toast').innerHTML).to.deep.equal('');
```
Comment (Vitalii-Misechko) [inline test/components/custom-toast-message.test.js:25]: `nit` should it be `.to.equal` instead?

File: components/custom-toast-message.js (hunk 1 above)
Comment (rohitvinnakota) [reply components/custom-toast-message.js:20]: My thinking was we can add additional toast messages as needed as functions and reuse the same component to conditionally render those messages; essentially calling those functions to trigger the toast message.

Comment (Vitalii-Misechko) [reply components/custom-toast-message.js:20]: sounds reasonable

**Review by Vitalii-Misechko — APPROVED** [review]

File: components/card-selection-list.js [hunk 5]
```diff
@@ -235,13 +235,13 @@ class CardSelectionList extends RtlMixin(Localizer(LitElement)) {
 
 	_handleThresholdFieldChange(event) {
 		const newValue = Number(event.target.value);
-		if (isNaN(newValue) || newValue < lastSysAccessThresholdMinDays || newValue > lastSysAccessThresholdMaxDays) {
-			return;
-		}
-
 		this.lastAccessThresholdDays = Math.floor(newValue);
```
Comment (bemailloux) [inline components/card-selection-list.js:238]: So what happens if there's an invalid input and the value is NaN? Does the input field show NaN?

(I think that's fine, but I also think the convention is for an input field to correct itself if there is an invalid input, which is something you've removed here.)

**Review by bemailloux — COMMENTED** [review]
Is there a corresponding server-side validation PR?

Also, what happens if we tried to save server-side and that failed? it doesn't look like there's error handling for that case.

File: components/card-selection-list.js (hunk 5 above)
Comment (rohitvinnakota) [reply components/card-selection-list.js:238]: When you click out, it makes the input field blank. I think it was the same functionality before.

**Comment by rohitvinnakota** [comment]
> Is there a corresponding server-side validation PR?
> 
> Also, what happens if we tried to save server-side and that failed? it doesn't look like there's error handling for that case.

Hmm yeah. I could add a quick check to return a failed response if we pass in an invalid value, but that should not happen with the client side validation.

**Comment by johngwilkinson** [comment]
> > Is there a corresponding server-side validation PR?
> > Also, what happens if we tried to save server-side and that failed? it doesn't look like there's error handling for that case.
> 
> Hmm yeah. I could add a quick check to return a failed response if we pass in an invalid value, but that should not happen with the client side validation.

We do need to show an error if the server returns 404 or 500 or the like, too - let's say the user's internet connection has dropped.

File: components/custom-toast-message.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+class CustomToastMessage extends RtlMixin(Localizer(LitElement)) {
+
+	render() {
+		return html`
+			<d2l-alert-toast type="critical"></d2l-alert-toast>
+		`;
+	}
+
+	systemLastAccessError() {
+		this.shadowRoot.querySelector('d2l-alert-toast').innerHTML = this.localize('settings:invalidSystemAccessValueToast');
```
Comment (bemailloux) [inline components/custom-toast-message.js:15]: You shouldn't use innerHTML to insert text like this: https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#Security_considerations

Instead `text` and `open` can just become properties in this class (basically what Vitalii said).

Comment (rohitvinnakota) [reply components/custom-toast-message.js]: I added `text ` as a prop, but added a function for open since that way, the message renders more than once.

**Review by bemailloux — APPROVED** [review]

**Pull request merged**